File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Camera/CameraSystem.lua
-- Compiled with unity-ts v2.1.0-75
local Signal = require("Shared/TS/Util/Signal").Signal
local Spring = require("Shared/TS/Util/Spring").Spring
local _Timer = require("Shared/TS/Util/Timer")
local OnLateUpdate = _Timer.OnLateUpdate
local OnUpdate = _Timer.OnUpdate
local StaticCameraMode = require("Client/TS/Controllers/Global/Camera/DefaultCameraModes/StaticCameraMode").StaticCameraMode
local TransitionMode
do
	TransitionMode = setmetatable({}, {
		__tostring = function()
			return "TransitionMode"
		end,
	})
	TransitionMode.__index = TransitionMode
	function TransitionMode.new(...)
		local self = setmetatable({}, TransitionMode)
		return self:constructor(...) or self
	end
	function TransitionMode:constructor(config, modeStart, modeGoal, onDone)
		self.config = config
		self.modeStart = modeStart
		self.modeGoal = modeGoal
		self.onDone = onDone
		self.start = Time.time
	end
	function TransitionMode:OnStart()
	end
	function TransitionMode:OnStop()
		self.modeStart:OnStop()
		self.modeGoal:OnStop()
	end
	function TransitionMode:OnUpdate(dt)
	end
	function TransitionMode:OnPostUpdate()
	end
	function TransitionMode:OnLateUpdate(dt)
		local transformStart = self.modeStart:OnLateUpdate(dt)
		local transformGoal = self.modeGoal:OnLateUpdate(dt)
		local alpha = math.clamp((Time.time - self.start) / self.config.Duration, 0, 1)
		if alpha == 1 then
			self.onDone()
		end
		return transformStart:Lerp(transformGoal, alpha)
	end
end
--[[
	*
	* Drives the camera modes.
]]
local CameraSystem
do
	CameraSystem = setmetatable({}, {
		__tostring = function()
			return "CameraSystem"
		end,
	})
	CameraSystem.__index = CameraSystem
	function CameraSystem.new(...)
		local self = setmetatable({}, CameraSystem)
		return self:constructor(...) or self
	end
	function CameraSystem:constructor(camera)
		self.camera = camera
		self.currentMode = StaticCameraMode.new(Vector3.new(0, 10, 0), Quaternion.identity)
		self.modeCleared = true
		self.fovSpringMoving = false
		self.fovSpringMovingStart = 0
		self.ModeChangedBegin = Signal.new()
		self.ModeChangedEnd = Signal.new()
		self.transform = camera.transform
		self.fovSpring = Spring.new(Vector3.new(camera.fieldOfView, 0, 0), 5)
		self.currentMode:OnStart(camera)
		local updateFov = function(dt)
			camera.fieldOfView = self.fovSpring:update(dt).x
			if Time.time - self.fovSpringMovingStart > 2 and math.abs(self.fovSpring.velocity.x) < 0.01 then
				self.fovSpringMoving = false
			end
		end
		OnUpdate:ConnectWithPriority(400, function(dt)
			self.currentMode:OnUpdate(dt)
		end)
		OnLateUpdate:ConnectWithPriority(0, function(dt)
			local camTransform = self.currentMode:OnLateUpdate(dt)
			self.transform:SetPositionAndRotation(camTransform.position, camTransform.rotation)
			self.currentMode:OnPostUpdate(self.camera)
			if self.fovSpringMoving then
				updateFov(dt)
			end
		end)
	end
	function CameraSystem:GetMode()
		return self.currentMode
	end
	function CameraSystem:SetMode(mode, transition)
		if mode == self.currentMode then
			return nil
		end
		self.modeCleared = false
		if transition == nil then
			local oldMode = self.currentMode
			self.ModeChangedBegin:Fire(mode, oldMode)
			self.currentMode:OnStop()
			self.currentMode = mode
			self.currentMode:OnStart(self.camera)
			self.ModeChangedEnd:Fire(mode, oldMode)
		else
			local oldMode = self.currentMode
			self.ModeChangedBegin:Fire(mode, oldMode)
			mode:OnStart(self.camera)
			self.currentMode = TransitionMode.new(transition, oldMode, mode, function()
				oldMode:OnStop()
				self.currentMode = mode
				self.ModeChangedEnd:Fire(mode, oldMode)
			end)
		end
	end
	function CameraSystem:ClearMode(transition)
		if self.onClearCallback then
			self:SetMode(self.onClearCallback(), transition)
		else
			self:SetMode(StaticCameraMode.new(self.transform.position, self.transform.rotation), transition)
			self.modeCleared = true
		end
	end
	function CameraSystem:SetOnClearCallback(onClearCallback)
		self.onClearCallback = onClearCallback
		if self.modeCleared and onClearCallback then
			self:SetMode(onClearCallback())
		end
	end
	function CameraSystem:SetFOV(fieldOfView, immediate)
		if immediate == nil then
			immediate = false
		end
		if immediate then
			self.fovSpring:resetTo(Vector3.new(fieldOfView, 0, 0))
			self.camera.fieldOfView = fieldOfView
			self.fovSpringMoving = false
		else
			self.fovSpring.goal = Vector3.new(fieldOfView, 0, 0)
			self.fovSpringMoving = true
			self.fovSpringMovingStart = Time.time
		end
	end
end
return {
	CameraSystem = CameraSystem,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/Camera/CameraSystem.ts
import { Signal, SignalPriority } from "Shared/Util/Signal";
import { Spring } from "Shared/Util/Spring";
import { OnLateUpdate, OnUpdate } from "Shared/Util/Timer";
import { CameraMode } from "./CameraMode";
import { CameraModeTransition } from "./CameraModeTransition";
import { StaticCameraMode } from "./DefaultCameraModes/StaticCameraMode";
import { CameraReferences } from "./CameraReferences";

class TransitionMode implements CameraMode {
	private readonly start: number;

	constructor(
		private readonly config: CameraModeTransition,
		private readonly modeStart: CameraMode,
		private readonly modeGoal: CameraMode,
		private readonly onDone: () => void,
	) {
		this.start = Time.time;
	}

	OnStart() {}

	OnStop() {
		this.modeStart.OnStop();
		this.modeGoal.OnStop();
	}

	OnUpdate(dt: number) {}

	OnPostUpdate() {}

	OnLateUpdate(dt: number) {
		const transformStart = this.modeStart.OnLateUpdate(dt);
		const transformGoal = this.modeGoal.OnLateUpdate(dt);
		const alpha = math.clamp((Time.time - this.start) / this.config.Duration, 0, 1);
		if (alpha === 1) {
			this.onDone();
		}
		return transformStart.Lerp(transformGoal, alpha);
	}
}

/**
 * Drives the camera modes.
 */
export class CameraSystem {
	private currentMode: CameraMode = new StaticCameraMode(new Vector3(0, 10, 0), Quaternion.identity);
	private readonly transform: Transform;
	private readonly camera: Camera;
	private readonly allCameras: Camera[];
	private onClearCallback?: () => CameraMode;
	private modeCleared = true;

	private readonly fovSpring: Spring;
	private fovSpringMoving = false;
	private fovSpringMovingStart = 0;

	public readonly ModeChangedBegin = new Signal<[newMode: CameraMode, oldMode: CameraMode]>();
	public readonly ModeChangedEnd = new Signal<[newMode: CameraMode, oldMode: CameraMode]>();

	constructor() {
		const ref = CameraReferences.Instance();
		this.camera = ref.mainCamera;
		this.allCameras = [ref.mainCamera, ref.uiCamera, ref.fpsCamera];
		this.transform = this.camera.transform;
		this.fovSpring = new Spring(new Vector3(this.camera.fieldOfView, 0, 0), 5);

		this.currentMode.OnStart(this.camera);

		OnUpdate.ConnectWithPriority(SignalPriority.LOWEST, (dt) => {
			this.currentMode.OnUpdate(dt);
		});

		OnLateUpdate.ConnectWithPriority(SignalPriority.HIGHEST, (dt) => {
			const camTransform = this.currentMode.OnLateUpdate(dt);
			this.transform.SetPositionAndRotation(camTransform.position, camTransform.rotation);
			this.currentMode.OnPostUpdate(this.camera);
			if (this.fovSpringMoving) {
				this.UpdateFOVSpring(dt);
			}
		});
	}

	/**
	 * Gets a reference to the current camera mode object.
	 * @returns Camera mode.
	 */
	public GetMode() {
		return this.currentMode;
	}

	/**
	 * Set the current camera mode. If `transition` is provided, then the new
	 * mode will be interpolated from the old mode based on the configuration
	 * provided within `transition`. Otherwise, the camera will snap immediately
	 * to the new mode.
	 *
	 * @param mode New mode.
	 * @param transition Optional transition configuration.
	 */
	public SetMode(mode: CameraMode, transition?: CameraModeTransition) {
		if (mode === this.currentMode) return;
		this.modeCleared = false;

		if (transition === undefined) {
			const oldMode = this.currentMode;
			this.ModeChangedBegin.Fire(mode, oldMode);
			this.currentMode.OnStop();
			this.currentMode = mode;
			this.currentMode.OnStart(this.camera);
			this.ModeChangedEnd.Fire(mode, oldMode);
		} else {
			const oldMode = this.currentMode;
			this.ModeChangedBegin.Fire(mode, oldMode);
			mode.OnStart(this.camera);
			this.currentMode = new TransitionMode(transition, oldMode, mode, () => {
				oldMode.OnStop();
				this.currentMode = mode;
				this.ModeChangedEnd.Fire(mode, oldMode);
			});
		}
	}

	/**
	 * Sets the camera to a static view.
	 *
	 * @param transition Optional transition configuration.
	 */
	public ClearMode(transition?: CameraModeTransition) {
		if (this.onClearCallback) {
			this.SetMode(this.onClearCallback(), transition);
		} else {
			this.SetMode(new StaticCameraMode(this.transform.position, this.transform.rotation), transition);
			this.modeCleared = true;
		}
	}

	/**
	 * Sets a callback function that is called when the camera mode is cleared. This
	 * is useful for defaulting the camera system to a specific camera mode. Only
	 * one callback can be set.
	 *
	 * Leaving out the `onClearCallback` parameter will clear the callback.
	 *
	 * @param onClearCallback Callback.
	 */
	public SetOnClearCallback(onClearCallback?: () => CameraMode) {
		this.onClearCallback = onClearCallback;
		if (this.modeCleared && onClearCallback) {
			this.SetMode(onClearCallback());
		}
	}

	/**
	 * Set the camera's field-of-view.
	 * @param fieldOfView Field of view.
	 * @param immediate If `true`, goes directly to the FOV without springing towards it.
	 */
	public SetFOV(fieldOfView: number, immediate = false) {
		if (immediate) {
			this.fovSpring.resetTo(new Vector3(fieldOfView, 0, 0));
			this.UpdateFOV(fieldOfView);
			this.fovSpringMoving = false;
		} else {
			this.fovSpring.goal = new Vector3(fieldOfView, 0, 0);
			this.fovSpringMoving = true;
			this.fovSpringMovingStart = Time.time;
		}
	}

	private UpdateFOVSpring(dt: number) {
		this.UpdateFOV(this.fovSpring.update(dt).x);
		if (Time.time - this.fovSpringMovingStart > 2 && math.abs(this.fovSpring.velocity.x) < 0.01) {
			this.fovSpringMoving = false;
		}
	}

	private UpdateFOV(newFOV: number) {
		for (let i = 0; i < this.allCameras.size(); i++) {
			this.allCameras[i].fieldOfView = newFOV;
		}
	}
}
